// node-pearls/dist/pipe.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

exports.default = function (tasks) {
	var onEnd;

	function run() {
		var task = tasks.shift();

		for (var _len = arguments.length, params = Array(_len), _key = 0; _key < _len; _key++) {
			params[_key] = arguments[_key];
		}

		if (task) {
			var isSync = true;

			var result = task.apply({
				async: function async() {
					isSync = false;
					return function () {
						run.apply(undefined, arguments);
					};
				}
			}, params);

			if (isSync) {
				run(result);
			}
		} else {
			onEnd.apply(undefined, params);
		}
	};

	return {
		start: function start() {
			for (var _len2 = arguments.length, params = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
				params[_key2] = arguments[_key2];
			}

			var listeners = [];
			var result;

			onEnd = function onEnd() {
				for (var _len3 = arguments.length, params = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
					params[_key3] = arguments[_key3];
				}

				result = params;

				listeners.forEach(function (listener) {
					listener.apply(undefined, params);
				});
			};

			run.apply(undefined, params);

			return {
				end: function end(listener) {
					if (result) {
						listener.apply(undefined, _toConsumableArray(result));
					} else {
						listeners.push(listener);
					}
					return this;
				}
			};
		}
	};
};

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } } /** 数据处理管道
                                                                                                                                                                                                     * @example
                                                                                                                                                                                                     	pipe(tasks)
                                                                                                                                                                                                     	.start(params)
                                                                                                                                                                                                     	.end(function(){
                                                                                                                                                                                                     		...
                                                                                                                                                                                                     	})
                                                                                                                                                                                                     * @example
                                                                                                                                                                                                     	pipe([
                                                                                                                                                                                                     		function(...params){
                                                                                                                                                                                                    			console.log(params);
                                                                                                                                                                                                    
                                                                                                                                                                                                    			return "abc";
                                                                                                                                                                                                     		},
                                                                                                                                                                                                     		function(...params){
                                                                                                                                                                                                     			console.log(params);
                                                                                                                                                                                                    
                                                                                                                                                                                                    			var callback = this.async();
                                                                                                                                                                                                    			setTimeout(function(){
                                                                                                                                                                                                    				callback(4, 5, 6);
                                                                                                                                                                                                    			}, 1000);
                                                                                                                                                                                                     		},
                                                                                                                                                                                                     		function(...params){
                                                                                                                                                                                                    			console.log(params);
                                                                                                                                                                                                    
                                                                                                                                                                                                    			return "complete";
                                                                                                                                                                                                     		}
                                                                                                                                                                                                     	])
                                                                                                                                                                                                     	.start(1, 2, 3)
                                                                                                                                                                                                     	.end(function(...params){
                                                                                                                                                                                                    		console.log(params);
                                                                                                                                                                                                     	});
                                                                                                                                                                                                     */


;
